cmd/kubectl-k8ssandra/tools/estimate.go

```diff
@@ -0,0 +1,171 @@
+package tools
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/pkg/errors"
+
+	"github.com/k8ssandra/k8ssandra-client/pkg/kubernetes"
+	"github.com/k8ssandra/k8ssandra-client/pkg/scheduler"
+	"github.com/spf13/cobra"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/cli-runtime/pkg/genericclioptions"
+)
+
+var (
+	estimateExample = `
+	# Estimate if pods will fit the cluster
+	%[1]s estimate [<args>]
+
+	# Estimate if 4 pods each with 2Gi of memory and 2 vCPUs will be able to run on the cluster.
+	# All CPU values and memory use Kubernetes notation
+	%[1]s estimate --count 4 --memory 2Gi --cpu 2000m
+	`
+	errInvalidCount = errors.New("Count of pods must be higher than 0")
+)
+
+type estimateOptions struct {
+	configFlags *genericclioptions.ConfigFlags
+	genericclioptions.IOStreams
+
+	count  int
+	memory string
+	cpu    string
+
+	cpuQuantity    resource.Quantity
+	memoryQuantity resource.Quantity
+}
+
+func newEstimateOptions(streams genericclioptions.IOStreams) *estimateOptions {
+	return &estimateOptions{
+		configFlags: genericclioptions.NewConfigFlags(true),
+		IOStreams:   streams,
+	}
+}
+
+// NewCmd provides a cobra command wrapping newAddOptions
+func NewEstimateCmd(streams genericclioptions.IOStreams) *cobra.Command {
+	o := newEstimateOptions(streams)
+
+	cmd := &cobra.Command{
+		Use:     "estimate [flags]",
+		Short:   "Estimate if datacenter can be expanded",
+		Example: fmt.Sprintf(estimateExample, "kubectl k8ssandra tools estimate"),
+		RunE: func(c *cobra.Command, args []string) error {
+			if err := o.Complete(c, args); err != nil {
+				return err
+			}
+			if err := o.Validate(); err != nil {
+				return err
+			}
+			if err := o.Run(); err != nil {
+				return err
```

Comment:
**Issue:** doing things this way always spits out the help text for the command in the console, along with the actual error:
```
k8ssandra-client % ./kubectl-k8ssandra tools estimate --count 25 --memory 2Gi --cpu 2000m 
Error: Unable to schedule the pods: unable to schedule all the pods, requested: 25, schedulable: 5
Usage:
  k8ssandra tools estimate [flags]

Examples:

        # Estimate if pods will fit the cluster
        kubectl k8ssandra tools estimate estimate [<args>]

        # Estimate if 4 pods each with 2Gi of memory and 2 vCPUs will be able to run on the cluster.
        # All CPU values and memory use Kubernetes notation
        kubectl k8ssandra tools estimate estimate --count 4 --memory 2Gi --cpu 2000m
        

Flags:
      --as string                      Username to impersonate for the operation. User could be a regular user or a service account in a namespace.
      --as-group stringArray           Group to impersonate for the operation, this flag can be repeated to specify multiple groups.
      --as-uid string                  UID to impersonate for the operation.
      --cache-dir string               Default cache directory (default "/Users/adejanovski/.kube/cache")
      --certificate-authority string   Path to a cert file for the certificate authority
      --client-certificate string      Path to a client certificate file for TLS
      --client-key string              Path to a client key file for TLS
      --cluster string                 The name of the kubeconfig cluster to use
      --context string                 The name of the kubeconfig context to use
      --count int                      new nodes to create
...
```

It makes it a little challenging to figure out what's going on.
Also we don't get a message if the pods can successfully be scheduled.

Reply:
Changed to this:

```console
➜  k8ssandra-client git:(add_scheduler_predictor) ✗ ./kubectl-k8ssandra tools estimate --count 6 --memory 2Gi --cpu 2000m    
2024/12/03 13:24:32 Pods can be scheduled to current cluster kind-kind
➜  k8ssandra-client git:(add_scheduler_predictor) ✗ ./kubectl-k8ssandra tools estimate --count 16 --memory 2Gi --cpu 2000m
2024/12/03 13:24:36 Error: Unable to schedule the pods to current cluster kind-kind: unable to schedule all the pods, requested: 16, schedulable: 6
```